File: Medium/210.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution {
  public int[] FindOrder(int numCourses, int[][] prerequisites) {
    Dictionary<int, List<int>> map = new Dictionary<int, List<int>>();
    List<int> res = new List<int>();
    HashSet<int> visited = new HashSet<int>();
    HashSet<int> currentPath = new HashSet<int>();
    bool hasCycle = false;

    // Build
    foreach (int[] pair in prerequisites) {
      int course = pair[0];
      int pre = pair[1];

      if (!map.ContainsKey(pre)) {
        map[pre] = new List<int>();
      }

      map[pre].Add(course);

      if (!map.ContainsKey(course)) {
        map[course] = new List<int>();
      }
    }

    void Dfs(int course) {
      if (currentPath.Contains(course)) {
        hasCycle = true;
        return;
      }

      if (visited.Contains(course) || hasCycle) {
        return;
      }

      // 防止重复访问
      // visited 和 currentPath 的区别：
      // visited 是全局访问过的节点
      // currentPath 是当前 DFS 路径上的节点
      visited.Add(course);
      currentPath.Add(course);

      // 如果 map 里没有这个 course，说明它没有后续课程
      // 对应 JS 里 map[course] = [] 的情况
      if (map.ContainsKey(course)) {
        foreach (int c in map[course]) {
          Dfs(c);

          if (hasCycle) {
            return;
          }
        }
      }

      // 严格按照 Topological Sort 的方式
      // 后序遍历的时候加上 current 节点
      res.Add(course);
      currentPath.Remove(course);
    }

    for (int i = 0; i < numCourses; i++) {
      Dfs(i);
    }

    if (hasCycle) {
      return new int[0];
    }

    res.Reverse();
    return res.ToArray();
  }
}

File: Medium/210.py
from typing import List
from collections import defaultdict


class Solution:
  def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
    map = defaultdict(list)
    res = []
    visited = set()
    current_path = set()
    has_cycle = False

    # Build
    for course, pre in prerequisites:
      map[pre].append(course)
      map[course]

    def dfs(course):
      nonlocal has_cycle

      if course in current_path:
        has_cycle = True
        return

      if course in visited or has_cycle:
        return

      # 防止重复访问
      # visited 和 current_path 的区别：
      # visited 是全局访问过的节点
      # current_path 是当前 DFS 路径上的节点
      visited.add(course)
      current_path.add(course)

      # map[course] = [] 情况下这一条不会运行
      # 前提是 map[course] 被初始化为 []
      for c in map[course]:
        dfs(c)

        if has_cycle:
          return

      # 严格按照 Topological Sort 的方式
      # 后序遍历的时候加上 current 节点
      res.append(course)
      current_path.remove(course)

    for i in range(numCourses):
      dfs(i)

    # 此乃 py 的三元表达式的写法
    return [] if has_cycle else res[::-1]